File: app/src/main/java/com/example/tdm_android/activities/FavouritesActivity.java
package com.example.tdm_android.activities;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import android.widget.Toolbar;

import com.example.tdm_android.R;
import com.example.tdm_android.adapters.CharacterAdapter;
import com.example.tdm_android.models.Character;
import com.google.android.material.navigation.NavigationView;

import java.util.ArrayList;
import java.util.List;

public class FavouritesActivity extends AppCompatActivity {

    DrawerLayout drawerLayout;
    NavigationView navigationView;
    Toolbar toolbar;
    ActionBarDrawerToggle actionBarDrawerToggle;

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {

        if(actionBarDrawerToggle.onOptionsItemSelected(item)) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_favourites);

        drawerLayout = findViewById(R.id.drawer_layout);
        navigationView = findViewById(R.id.navigationView);
        actionBarDrawerToggle = new ActionBarDrawerToggle(this, drawerLayout, R.string.menu_open, R.string.menu_close);
        drawerLayout.addDrawerListener(actionBarDrawerToggle);
        actionBarDrawerToggle.syncState();
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        navigationView.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {

                switch (item.getItemId()) {
                    case R.id.nav_home:
                        drawerLayout.closeDrawer(GravityCompat.START);
                        Intent intent = new Intent(FavouritesActivity.this, FilterActivity.class);
                        startActivity(intent);
                        break;
                    case R.id.nav_favourites:
                        drawerLayout.closeDrawer(GravityCompat.START);
                        Intent intent1 = new Intent(FavouritesActivity.this, FavouritesActivity.class);
                        startActivity(intent1);
                        break;
                }

                return true;
            }
        });

        setupAdapter();
    }

    private void setupAdapter() {

        RecyclerView rvCharacters = findViewById(R.id.listRecyclerView);

        CharacterAdapter charactersAdapter = new CharacterAdapter(getCharacters(), character -> {
            Intent intent = new Intent(FavouritesActivity.this, DetailActivity.class);
            startActivity(intent);
        });

        rvCharacters.setLayoutManager(new GridLayoutManager(this, 2));
        rvCharacters.setAdapter(charactersAdapter);
    }


    private List<Character> getCharacters() {
        List<Character> listCharacters = new ArrayList<>();
        listCharacters.add(new Character("Arya Stark", "Northmen"));
        listCharacters.add(new Character("Arya Stark", "Northmen"));
        listCharacters.add(new Character("Arya Stark", "Northmen"));
        listCharacters.add(new Character("Arya Stark", "Northmen"));

        return listCharacters;
    }
}

File: app/src/main/java/com/example/tdm_android/activities/FavouritesActivity.kt
package com.example.tdm_android.activities

import androidx.recyclerview.widget.RecyclerView
import com.example.tdm_android.R
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.drawerlayout.widget.DrawerLayout
import com.google.android.material.navigation.NavigationView
import androidx.appcompat.app.ActionBarDrawerToggle
import android.content.Intent
import com.example.tdm_android.adapters.CharacterAdapter
import androidx.recyclerview.widget.GridLayoutManager
import android.view.MenuItem
import android.widget.TextView
import androidx.lifecycle.lifecycleScope
import com.example.tdm_android.functions.triggerByChoosingNavigationMenuItem
import com.example.tdm_android.managers.FavouriteCharacterManager
import com.example.tdm_android.models.Character

class FavouritesActivity : AppCompatActivity() {

    lateinit var drawerLayout: DrawerLayout
    lateinit var navigationView: NavigationView
    private lateinit var actionBarDrawerToggle: ActionBarDrawerToggle
    private lateinit var tvDescriptionScreenFavourites: TextView

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (actionBarDrawerToggle.onOptionsItemSelected(item)) {
            true
        } else super.onOptionsItemSelected(item)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_favourites)

        drawerLayout = findViewById(R.id.drawer_layout)
        navigationView = findViewById(R.id.navigationView)
        tvDescriptionScreenFavourites = findViewById(R.id.tvDescriptionScreenFavourites)
        actionBarDrawerToggle = ActionBarDrawerToggle(this, drawerLayout, R.string.menu_open, R.string.menu_close)
        drawerLayout.addDrawerListener(actionBarDrawerToggle)
        actionBarDrawerToggle.syncState()
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        triggerByChoosingNavigationMenuItem(lifecycleScope, navigationView, drawerLayout)

        setupAdapter()
    }

    private fun setupAdapter() {
        val rvCharacters = findViewById<RecyclerView>(R.id.listRecyclerView)

        val listCharacters: MutableList<Character> = ArrayList()

        FavouriteCharacterManager.getInstance(this@FavouritesActivity)?.favouriteCharacters?.forEach { favouriteCharacter ->
            if (favouriteCharacter != null) {
                listCharacters.add(Character(favouriteCharacter.name, favouriteCharacter.gender, url = favouriteCharacter.id.toString()))
            }
        }

        if(listCharacters.size === 0) tvDescriptionScreenFavourites.text = "Favourites - Empty"

        val charactersAdapter = CharacterAdapter(listCharacters) { character: Character? ->
            Intent(this@FavouritesActivity, DetailActivity::class.java).also {
                if (character != null) {
                    it.putExtra("id", character.url?.filter { it -> it.isDigit() })
                }
                startActivity(it)
            }
        }
        rvCharacters.layoutManager = GridLayoutManager(this, 2)
        rvCharacters.adapter = charactersAdapter
    }

}